File: mobile/API/SupportAPI.swift
//
//  SupportAPI.swift
//  ALADDIN Mobile Security
//
//  Support API Integration for iOS
//  API client for Super AI Support Assistant and Psychological Support Agent
//
//  Created by ALADDIN Security Team
//  Date: 2025-01-27
//  Version: 1.0
//

import Foundation
import Combine

// MARK: - Support API Models
struct SupportRequest: Codable {
    let question: String
    let userId: String
    let context: SupportContext
    let category: SupportCategory?
    let language: String
    let timestamp: Date
    
    init(question: String, userId: String, context: SupportContext, category: SupportCategory? = nil, language: String = "ru") {
        self.question = question
        self.userId = userId
        self.context = context
        self.category = category
        self.language = language
        self.timestamp = Date()
    }
}

struct SupportContext: Codable {
    let currentScreen: String?
    let userAgeGroup: String?
    let emotionalState: String?
    let deviceInfo: DeviceInfo
    let sessionId: String
    
    init(currentScreen: String? = nil, userAgeGroup: String? = nil, emotionalState: String? = nil, deviceInfo: DeviceInfo, sessionId: String) {
        self.currentScreen = currentScreen
        self.userAgeGroup = userAgeGroup
        self.emotionalState = emotionalState
        self.deviceInfo = deviceInfo
        self.sessionId = sessionId
    }
}

struct DeviceInfo: Codable {
    let platform: String
    let version: String
    let model: String
    let screenSize: String
    let orientation: String
    
    init(platform: String = "iOS", version: String, model: String, screenSize: String, orientation: String = "portrait") {
        self.platform = platform
        self.version = version
        self.model = model
        self.screenSize = screenSize
        self.orientation = orientation
    }
}

struct SupportResponse: Codable {
    let answer: String
    let category: String
    let suggestedActions: [String]
    let relatedTopics: [String]
    let confidence: Double
    let responseTime: TimeInterval
    let timestamp: Date
    let sessionId: String
    let followUpQuestions: [String]?
    let emotionalAnalysis: EmotionalAnalysis?
    let priority: SupportPriority
}

struct EmotionalAnalysis: Codable {
    let emotion: String
    let confidence: Double
    let sentiment: String
    let recommendations: [String]
}

enum SupportCategory: String, CaseIterable, Codable {
    case cybersecurity = "cybersecurity"
    case familySupport = "family_support"
    case medicalSupport = "medical_support"
    case education = "education"
    case finance = "finance"
    case household = "household"
    case psychology = "psychology"
    case technology = "technology"
    case legal = "legal"
    case travel = "travel"
    case entertainment = "entertainment"
    case health = "health"
    case fitness = "fitness"
    case relationships = "relationships"
    case career = "career"
    case business = "business"
    case shopping = "shopping"
    case cooking = "cooking"
    case gardening = "gardening"
    case repair = "repair"
}

enum SupportPriority: String, CaseIterable, Codable {
    case low = "low"
    case medium = "medium"
    case high = "high"
    case urgent = "urgent"
}

// MARK: - Support API Client
class SupportAPIClient: ObservableObject {
    
    // MARK: - Properties
    private let baseURL: String
    private let session: URLSession
    private let apiKey: String
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Initialization
    init(baseURL: String = "https://api.aladdin-security.com", apiKey: String) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        self.session = URLSession(configuration: config)
    }
    
    // MARK: - Public Methods
    
    /// Ask a question to the Super AI Support Assistant
    func askQuestion(_ request: SupportRequest) -> AnyPublisher<SupportResponse, Error> {
        let endpoint = "/api/v1/support/ask"
        return performRequest(endpoint: endpoint, method: "POST", body: request)
    }
    
    /// Get psychological support
    func getPsychologicalSupport(_ request: SupportRequest) -> AnyPublisher<SupportResponse, Error> {
        let endpoint = "/api/v1/support/psychological"
        return performRequest(endpoint: endpoint, method: "POST", body: request)
    }
    
    /// Get navigation help
    func getNavigationHelp(_ request: SupportRequest) -> AnyPublisher<SupportResponse, Error> {
        let endpoint = "/api/v1/support/navigation"
        return performRequest(endpoint: endpoint, method: "POST", body: request)
    }
    
    /// Get FAQ
    func getFAQ(category: SupportCategory? = nil) -> AnyPublisher<[FAQItem], Error> {
        var endpoint = "/api/v1/support/faq"
        if let category = category {
            endpoint += "?category=\(category.rawValue)"
        }
        return performRequest(endpoint: endpoint, method: "GET", body: nil as String?)
    }
    
    /// Get support categories
    func getSupportCategories() -> AnyPublisher<[SupportCategory], Error> {
        let endpoint = "/api/v1/support/categories"
        return performRequest(endpoint: endpoint, method: "GET", body: nil as String?)
    }
    
    /// Submit feedback
    func submitFeedback(_ feedback: SupportFeedback) -> AnyPublisher<SupportResponse, Error> {
        let endpoint = "/api/v1/support/feedback"
        return performRequest(endpoint: endpoint, method: "POST", body: feedback)
    }
    
    /// Get support history
    func getSupportHistory(userId: String, limit: Int = 50) -> AnyPublisher<[SupportHistoryItem], Error> {
        let endpoint = "/api/v1/support/history?userId=\(userId)&limit=\(limit)"
        return performRequest(endpoint: endpoint, method: "GET", body: nil as String?)
    }
    
    // MARK: - Private Methods
    
    private func performRequest<T: Codable, U: Codable>(
        endpoint: String,
        method: String,
        body: T?
    ) -> AnyPublisher<U, Error> {
        guard let url = URL(string: baseURL + endpoint) else {
            return Fail(error: APIError.invalidURL)
                .eraseToAnyPublisher()
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("iOS", forHTTPHeaderField: "X-Platform")
        request.setValue(Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0", forHTTPHeaderField: "X-App-Version")
        
        if let body = body {
            do {
                request.httpBody = try JSONEncoder().encode(body)
            } catch {
                return Fail(error: APIError.encodingError(error))
                    .eraseToAnyPublisher()
            }
        }
        
        return session.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: U.self, decoder: JSONDecoder())
            .receive(on: DispatchQueue.main)
            .eraseToAnyPublisher()
    }
}

// MARK: - Additional Models
struct FAQItem: Codable {
    let id: String
    let question: String
    let answer: String
    let category: String
    let tags: [String]
    let helpful: Int
    let notHelpful: Int
    let lastUpdated: Date
}

struct SupportFeedback: Codable {
    let sessionId: String
    let rating: Int
    let comment: String?
    let helpful: Bool
    let category: String
    let timestamp: Date
}

struct SupportHistoryItem: Codable {
    let id: String
    let question: String
    let answer: String
    let category: String
    let timestamp: Date
    let rating: Int?
    let helpful: Bool?
}

// MARK: - API Errors
enum APIError: Error, LocalizedError {
    case invalidURL
    case noData
    case decodingError(Error)
    case encodingError(Error)
    case networkError(Error)
    case serverError(Int, String)
    case unauthorized
    case forbidden
    case notFound
    case rateLimited
    case timeout
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .noData:
            return "No data received"
        case .decodingError(let error):
            return "Decoding error: \(error.localizedDescription)"
        case .encodingError(let error):
            return "Encoding error: \(error.localizedDescription)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .serverError(let code, let message):
            return "Server error \(code): \(message)"
        case .unauthorized:
            return "Unauthorized access"
        case .forbidden:
            return "Access forbidden"
        case .notFound:
            return "Resource not found"
        case .rateLimited:
            return "Rate limit exceeded"
        case .timeout:
            return "Request timeout"
        }
    }
}

// MARK: - Support API Manager
class SupportAPIManager: ObservableObject {
    
    // MARK: - Properties
    private let apiClient: SupportAPIClient
    private let userId: String
    private let sessionId: String
    private var cancellables = Set<AnyCancellable>()
    
    @Published var isLoading = false
    @Published var lastError: APIError?
    @Published var supportHistory: [SupportHistoryItem] = []
    @Published var faqItems: [FAQItem] = []
    
    // MARK: - Initialization
    init(apiClient: SupportAPIClient, userId: String, sessionId: String = UUID().uuidString) {
        self.apiClient = apiClient
        self.userId = userId
        self.sessionId = sessionId
    }
    
    // MARK: - Public Methods
    
    /// Ask a question with automatic error handling
    func askQuestion(
        _ question: String,
        category: SupportCategory? = nil,
        context: SupportContext? = nil
    ) -> AnyPublisher<SupportResponse, Never> {
        isLoading = true
        lastError = nil
        
        let defaultContext = context ?? createDefaultContext()
        let request = SupportRequest(
            question: question,
            userId: userId,
            context: defaultContext,
            category: category
        )
        
        return apiClient.askQuestion(request)
            .handleEvents(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.lastError = error as? APIError
                    }
                }
            )
            .catch { error in
                Just(SupportResponse(
                    answer: "Извините, произошла ошибка при обработке вашего запроса. Пожалуйста, попробуйте еще раз.",
                    category: "error",
                    suggestedActions: ["Попробовать еще раз", "Обратиться в поддержку"],
                    relatedTopics: [],
                    confidence: 0.0,
                    responseTime: 0.0,
                    timestamp: Date(),
                    sessionId: self.sessionId,
                    followUpQuestions: nil,
                    emotionalAnalysis: nil,
                    priority: .low
                ))
            }
            .eraseToAnyPublisher()
    }
    
    /// Get psychological support
    func getPsychologicalSupport(
        _ question: String,
        ageGroup: String? = nil,
        emotionalState: String? = nil
    ) -> AnyPublisher<SupportResponse, Never> {
        isLoading = true
        lastError = nil
        
        let context = SupportContext(
            currentScreen: nil,
            userAgeGroup: ageGroup,
            emotionalState: emotionalState,
            deviceInfo: createDeviceInfo(),
            sessionId: sessionId
        )
        
        let request = SupportRequest(
            question: question,
            userId: userId,
            context: context,
            category: .psychology
        )
        
        return apiClient.getPsychologicalSupport(request)
            .handleEvents(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.lastError = error as? APIError
                    }
                }
            )
            .catch { error in
                Just(SupportResponse(
                    answer: "Извините, психологическая поддержка временно недоступна. Пожалуйста, попробуйте позже.",
                    category: "psychology",
                    suggestedActions: ["Попробовать позже", "Обратиться к специалисту"],
                    relatedTopics: [],
                    confidence: 0.0,
                    responseTime: 0.0,
                    timestamp: Date(),
                    sessionId: self.sessionId,
                    followUpQuestions: nil,
                    emotionalAnalysis: nil,
                    priority: .medium
                ))
            }
            .eraseToAnyPublisher()
    }
    
    /// Load FAQ items
    func loadFAQ(category: SupportCategory? = nil) {
        apiClient.getFAQ(category: category)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.lastError = error as? APIError
                    }
                },
                receiveValue: { [weak self] items in
                    self?.faqItems = items
                }
            )
            .store(in: &cancellables)
    }
    
    /// Load support history
    func loadSupportHistory() {
        apiClient.getSupportHistory(userId: userId)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.lastError = error as? APIError
                    }
                },
                receiveValue: { [weak self] items in
                    self?.supportHistory = items
                }
            )
            .store(in: &cancellables)
    }
    
    /// Submit feedback
    func submitFeedback(
        sessionId: String,
        rating: Int,
        comment: String? = nil,
        helpful: Bool,
        category: String
    ) {
        let feedback = SupportFeedback(
            sessionId: sessionId,
            rating: rating,
            comment: comment,
            helpful: helpful,
            category: category,
            timestamp: Date()
        )
        
        apiClient.submitFeedback(feedback)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.lastError = error as? APIError
                    }
                },
                receiveValue: { _ in
                    // Feedback submitted successfully
                }
            )
            .store(in: &cancellables)
    }
    
    // MARK: - Private Methods
    
    private func createDefaultContext() -> SupportContext {
        return SupportContext(
            currentScreen: nil,
            userAgeGroup: nil,
            emotionalState: nil,
            deviceInfo: createDeviceInfo(),
            sessionId: sessionId
        )
    }
    
    private func createDeviceInfo() -> DeviceInfo {
        let screenSize = UIScreen.main.bounds.size
        let screenSizeString = "\(Int(screenSize.width))x\(Int(screenSize.height))"
        
        return DeviceInfo(
            platform: "iOS",
            version: UIDevice.current.systemVersion,
            model: UIDevice.current.model,
            screenSize: screenSizeString,
            orientation: UIDevice.current.orientation.isPortrait ? "portrait" : "landscape"
        )
    }
}

// MARK: - Support API Extensions
extension SupportAPIManager {
    
    /// Quick question method for simple queries
    func quickQuestion(_ question: String) -> AnyPublisher<SupportResponse, Never> {
        return askQuestion(question)
    }
    
    /// Get help for specific screen
    func getScreenHelp(screen: String) -> AnyPublisher<SupportResponse, Never> {
        let context = SupportContext(
            currentScreen: screen,
            userAgeGroup: nil,
            emotionalState: nil,
            deviceInfo: createDeviceInfo(),
            sessionId: sessionId
        )
        
        let request = SupportRequest(
            question: "Помощь по экрану \(screen)",
            userId: userId,
            context: context,
            category: .technology
        )
        
        return apiClient.getNavigationHelp(request)
            .catch { error in
                Just(SupportResponse(
                    answer: "Помощь по экрану \(screen) временно недоступна.",
                    category: "technology",
                    suggestedActions: ["Попробовать позже"],
                    relatedTopics: [],
                    confidence: 0.0,
                    responseTime: 0.0,
                    timestamp: Date(),
                    sessionId: self.sessionId,
                    followUpQuestions: nil,
                    emotionalAnalysis: nil,
                    priority: .low
                ))
            }
            .eraseToAnyPublisher()
    }
}

File: mobile/API/SupportAPI.kt
/*
 * SupportAPI.kt
 * ALADDIN Mobile Security
 *
 * Support API Integration for Android
 * API client for Super AI Support Assistant and Psychological Support Agent
 *
 * Created by ALADDIN Security Team
 * Date: 2025-01-27
 * Version: 1.0
 */

package com.aladdin.security.api.support

import android.content.Context
import android.os.Build
import android.util.Log
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.*

// MARK: - Support API Models
data class SupportRequest(
    @SerializedName("question") val question: String,
    @SerializedName("user_id") val userId: String,
    @SerializedName("context") val context: SupportContext,
    @SerializedName("category") val category: String?,
    @SerializedName("language") val language: String,
    @SerializedName("timestamp") val timestamp: Date
) {
    constructor(
        question: String,
        userId: String,
        context: SupportContext,
        category: SupportCategory? = null,
        language: String = "ru"
    ) : this(
        question = question,
        userId = userId,
        context = context,
        category = category?.value,
        language = language,
        timestamp = Date()
    )
}

data class SupportContext(
    @SerializedName("current_screen") val currentScreen: String?,
    @SerializedName("user_age_group") val userAgeGroup: String?,
    @SerializedName("emotional_state") val emotionalState: String?,
    @SerializedName("device_info") val deviceInfo: DeviceInfo,
    @SerializedName("session_id") val sessionId: String
)

data class DeviceInfo(
    @SerializedName("platform") val platform: String,
    @SerializedName("version") val version: String,
    @SerializedName("model") val model: String,
    @SerializedName("screen_size") val screenSize: String,
    @SerializedName("orientation") val orientation: String
)

data class SupportResponse(
    @SerializedName("answer") val answer: String,
    @SerializedName("category") val category: String,
    @SerializedName("suggested_actions") val suggestedActions: List<String>,
    @SerializedName("related_topics") val relatedTopics: List<String>,
    @SerializedName("confidence") val confidence: Double,
    @SerializedName("response_time") val responseTime: Double,
    @SerializedName("timestamp") val timestamp: Date,
    @SerializedName("session_id") val sessionId: String,
    @SerializedName("follow_up_questions") val followUpQuestions: List<String>?,
    @SerializedName("emotional_analysis") val emotionalAnalysis: EmotionalAnalysis?,
    @SerializedName("priority") val priority: String
)

data class EmotionalAnalysis(
    @SerializedName("emotion") val emotion: String,
    @SerializedName("confidence") val confidence: Double,
    @SerializedName("sentiment") val sentiment: String,
    @SerializedName("recommendations") val recommendations: List<String>
)

enum class SupportCategory(val value: String) {
    CYBERSECURITY("cybersecurity"),
    FAMILY_SUPPORT("family_support"),
    MEDICAL_SUPPORT("medical_support"),
    EDUCATION("education"),
    FINANCE("finance"),
    HOUSEHOLD("household"),
    PSYCHOLOGY("psychology"),
    TECHNOLOGY("technology"),
    LEGAL("legal"),
    TRAVEL("travel"),
    ENTERTAINMENT("entertainment"),
    HEALTH("health"),
    FITNESS("fitness"),
    RELATIONSHIPS("relationships"),
    CAREER("career"),
    BUSINESS("business"),
    SHOPPING("shopping"),
    COOKING("cooking"),
    GARDENING("gardening"),
    REPAIR("repair")
}

enum class SupportPriority(val value: String) {
    LOW("low"),
    MEDIUM("medium"),
    HIGH("high"),
    URGENT("urgent")
}

// MARK: - Additional Models
data class FAQItem(
    @SerializedName("id") val id: String,
    @SerializedName("question") val question: String,
    @SerializedName("answer") val answer: String,
    @SerializedName("category") val category: String,
    @SerializedName("tags") val tags: List<String>,
    @SerializedName("helpful") val helpful: Int,
    @SerializedName("not_helpful") val notHelpful: Int,
    @SerializedName("last_updated") val lastUpdated: Date
)

data class SupportFeedback(
    @SerializedName("session_id") val sessionId: String,
    @SerializedName("rating") val rating: Int,
    @SerializedName("comment") val comment: String?,
    @SerializedName("helpful") val helpful: Boolean,
    @SerializedName("category") val category: String,
    @SerializedName("timestamp") val timestamp: Date
)

data class SupportHistoryItem(
    @SerializedName("id") val id: String,
    @SerializedName("question") val question: String,
    @SerializedName("answer") val answer: String,
    @SerializedName("category") val category: String,
    @SerializedName("timestamp") val timestamp: Date,
    @SerializedName("rating") val rating: Int?,
    @SerializedName("helpful") val helpful: Boolean?
)

// MARK: - API Errors
sealed class APIError : Exception() {
    object InvalidURL : APIError()
    object NoData : APIError()
    data class DecodingError(val error: Throwable) : APIError()
    data class EncodingError(val error: Throwable) : APIError()
    data class NetworkError(val error: Throwable) : APIError()
    data class ServerError(val code: Int, val message: String) : APIError()
    object Unauthorized : APIError()
    object Forbidden : APIError()
    object NotFound : APIError()
    object RateLimited : APIError()
    object Timeout : APIError()
}

// MARK: - Support API Client
class SupportAPIClient(
    private val baseURL: String = "https://api.aladdin-security.com",
    private val apiKey: String
) {
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(60, java.util.concurrent.TimeUnit.SECONDS)
        .writeTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .build()
    
    private val gson = Gson()
    private val jsonMediaType = "application/json".toMediaType()
    
    // MARK: - Public Methods
    
    suspend fun askQuestion(request: SupportRequest): Result<SupportResponse> {
        return performRequest("/api/v1/support/ask", "POST", request)
    }
    
    suspend fun getPsychologicalSupport(request: SupportRequest): Result<SupportResponse> {
        return performRequest("/api/v1/support/psychological", "POST", request)
    }
    
    suspend fun getNavigationHelp(request: SupportRequest): Result<SupportResponse> {
        return performRequest("/api/v1/support/navigation", "POST", request)
    }
    
    suspend fun getFAQ(category: SupportCategory? = null): Result<List<FAQItem>> {
        var endpoint = "/api/v1/support/faq"
        if (category != null) {
            endpoint += "?category=${category.value}"
        }
        return performRequest(endpoint, "GET", null)
    }
    
    suspend fun getSupportCategories(): Result<List<SupportCategory>> {
        return performRequest("/api/v1/support/categories", "GET", null)
    }
    
    suspend fun submitFeedback(feedback: SupportFeedback): Result<SupportResponse> {
        return performRequest("/api/v1/support/feedback", "POST", feedback)
    }
    
    suspend fun getSupportHistory(userId: String, limit: Int = 50): Result<List<SupportHistoryItem>> {
        val endpoint = "/api/v1/support/history?userId=$userId&limit=$limit"
        return performRequest(endpoint, "GET", null)
    }
    
    // MARK: - Private Methods
    
    private suspend fun <T> performRequest(
        endpoint: String,
        method: String,
        body: T?
    ): Result<SupportResponse> = withContext(Dispatchers.IO) {
        try {
            val url = "$baseURL$endpoint"
            val requestBuilder = Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("Authorization", "Bearer $apiKey")
                .addHeader("X-Platform", "Android")
                .addHeader("X-App-Version", "1.0")
            
            when (method) {
                "POST" -> {
                    val jsonBody = if (body != null) {
                        gson.toJson(body).toRequestBody(jsonMediaType)
                    } else {
                        "{}".toRequestBody(jsonMediaType)
                    }
                    requestBuilder.post(jsonBody)
                }
                "GET" -> requestBuilder.get()
                else -> throw APIError.InvalidURL
            }
            
            val request = requestBuilder.build()
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                if (responseBody != null) {
                    val result = gson.fromJson(responseBody, SupportResponse::class.java)
                    Result.success(result)
                } else {
                    Result.failure(APIError.NoData)
                }
            } else {
                when (response.code) {
                    401 -> Result.failure(APIError.Unauthorized)
                    403 -> Result.failure(APIError.Forbidden)
                    404 -> Result.failure(APIError.NotFound)
                    429 -> Result.failure(APIError.RateLimited)
                    else -> Result.failure(APIError.ServerError(response.code, response.message))
                }
            }
        } catch (e: IOException) {
            Result.failure(APIError.NetworkError(e))
        } catch (e: Exception) {
            Result.failure(APIError.DecodingError(e))
        }
    }
    
    private suspend fun <T> performRequest(
        endpoint: String,
        method: String,
        body: T?,
        responseType: Class<*>
    ): Result<T> = withContext(Dispatchers.IO) {
        try {
            val url = "$baseURL$endpoint"
            val requestBuilder = Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("Authorization", "Bearer $apiKey")
                .addHeader("X-Platform", "Android")
                .addHeader("X-App-Version", "1.0")
            
            when (method) {
                "POST" -> {
                    val jsonBody = if (body != null) {
                        gson.toJson(body).toRequestBody(jsonMediaType)
                    } else {
                        "{}".toRequestBody(jsonMediaType)
                    }
                    requestBuilder.post(jsonBody)
                }
                "GET" -> requestBuilder.get()
                else -> throw APIError.InvalidURL
            }
            
            val request = requestBuilder.build()
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                if (responseBody != null) {
                    val result = gson.fromJson(responseBody, responseType) as T
                    Result.success(result)
                } else {
                    Result.failure(APIError.NoData)
                }
            } else {
                when (response.code) {
                    401 -> Result.failure(APIError.Unauthorized)
                    403 -> Result.failure(APIError.Forbidden)
                    404 -> Result.failure(APIError.NotFound)
                    429 -> Result.failure(APIError.RateLimited)
                    else -> Result.failure(APIError.ServerError(response.code, response.message))
                }
            }
        } catch (e: IOException) {
            Result.failure(APIError.NetworkError(e))
        } catch (e: Exception) {
            Result.failure(APIError.DecodingError(e))
        }
    }
}

// MARK: - Support API Manager
class SupportAPIManager(
    private val context: Context,
    private val apiClient: SupportAPIClient,
    private val userId: String,
    private val sessionId: String = UUID().randomUUID().toString()
) {
    
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    
    var isLoading = false
    var lastError: APIError? = null
    var supportHistory: List<SupportHistoryItem> = emptyList()
    var faqItems: List<FAQItem> = emptyList()
    
    // MARK: - Public Methods
    
    fun askQuestion(
        question: String,
        category: SupportCategory? = null,
        context: SupportContext? = null,
        onSuccess: (SupportResponse) -> Unit,
        onError: (APIError) -> Unit
    ) {
        scope.launch {
            isLoading = true
            lastError = null
            
            val defaultContext = context ?: createDefaultContext()
            val request = SupportRequest(
                question = question,
                userId = userId,
                context = defaultContext,
                category = category
            )
            
            when (val result = apiClient.askQuestion(request)) {
                is Result.success -> {
                    isLoading = false
                    onSuccess(result.getOrThrow())
                }
                is Result.failure -> {
                    isLoading = false
                    lastError = result.exceptionOrNull() as? APIError ?: APIError.NetworkError(Exception("Unknown error"))
                    onError(lastError!!)
                }
            }
        }
    }
    
    fun getPsychologicalSupport(
        question: String,
        ageGroup: String? = null,
        emotionalState: String? = null,
        onSuccess: (SupportResponse) -> Unit,
        onError: (APIError) -> Unit
    ) {
        scope.launch {
            isLoading = true
            lastError = null
            
            val context = SupportContext(
                currentScreen = null,
                userAgeGroup = ageGroup,
                emotionalState = emotionalState,
                deviceInfo = createDeviceInfo(),
                sessionId = sessionId
            )
            
            val request = SupportRequest(
                question = question,
                userId = userId,
                context = context,
                category = SupportCategory.PSYCHOLOGY
            )
            
            when (val result = apiClient.getPsychologicalSupport(request)) {
                is Result.success -> {
                    isLoading = false
                    onSuccess(result.getOrThrow())
                }
                is Result.failure -> {
                    isLoading = false
                    lastError = result.exceptionOrNull() as? APIError ?: APIError.NetworkError(Exception("Unknown error"))
                    onError(lastError!!)
                }
            }
        }
    }
    
    fun loadFAQ(category: SupportCategory? = null) {
        scope.launch {
            when (val result = apiClient.getFAQ(category)) {
                is Result.success -> {
                    faqItems = result.getOrThrow()
                }
                is Result.failure -> {
                    lastError = result.exceptionOrNull() as? APIError ?: APIError.NetworkError(Exception("Unknown error"))
                }
            }
        }
    }
    
    fun loadSupportHistory() {
        scope.launch {
            when (val result = apiClient.getSupportHistory(userId)) {
                is Result.success -> {
                    supportHistory = result.getOrThrow()
                }
                is Result.failure -> {
                    lastError = result.exceptionOrNull() as? APIError ?: APIError.NetworkError(Exception("Unknown error"))
                }
            }
        }
    }
    
    fun submitFeedback(
        sessionId: String,
        rating: Int,
        comment: String? = null,
        helpful: Boolean,
        category: String
    ) {
        scope.launch {
            val feedback = SupportFeedback(
                sessionId = sessionId,
                rating = rating,
                comment = comment,
                helpful = helpful,
                category = category,
                timestamp = Date()
            )
            
            when (val result = apiClient.submitFeedback(feedback)) {
                is Result.success -> {
                    // Feedback submitted successfully
                }
                is Result.failure -> {
                    lastError = result.exceptionOrNull() as? APIError ?: APIError.NetworkError(Exception("Unknown error"))
                }
            }
        }
    }
    
    fun getScreenHelp(screen: String, onSuccess: (SupportResponse) -> Unit, onError: (APIError) -> Unit) {
        val context = SupportContext(
            currentScreen = screen,
            userAgeGroup = null,
            emotionalState = null,
            deviceInfo = createDeviceInfo(),
            sessionId = sessionId
        )
        
        val request = SupportRequest(
            question = "Помощь по экрану $screen",
            userId = userId,
            context = context,
            category = SupportCategory.TECHNOLOGY
        )
        
        scope.launch {
            when (val result = apiClient.getNavigationHelp(request)) {
                is Result.success -> {
                    onSuccess(result.getOrThrow())
                }
                is Result.failure -> {
                    lastError = result.exceptionOrNull() as? APIError ?: APIError.NetworkError(Exception("Unknown error"))
                    onError(lastError!!)
                }
            }
        }
    }
    
    // MARK: - Private Methods
    
    private fun createDefaultContext(): SupportContext {
        return SupportContext(
            currentScreen = null,
            userAgeGroup = null,
            emotionalState = null,
            deviceInfo = createDeviceInfo(),
            sessionId = sessionId
        )
    }
    
    private fun createDeviceInfo(): DeviceInfo {
        val displayMetrics = context.resources.displayMetrics
        val screenSize = "${displayMetrics.widthPixels}x${displayMetrics.heightPixels}"
        
        return DeviceInfo(
            platform = "Android",
            version = Build.VERSION.RELEASE,
            model = Build.MODEL,
            screenSize = screenSize,
            orientation = if (displayMetrics.widthPixels > displayMetrics.heightPixels) "landscape" else "portrait"
        )
    }
    
    fun cleanup() {
        scope.cancel()
    }
}

// MARK: - Flow-based API (Alternative approach)
class SupportAPIFlow(
    private val apiClient: SupportAPIClient,
    private val userId: String,
    private val sessionId: String = UUID().randomUUID().toString()
) {
    
    fun askQuestionFlow(
        question: String,
        category: SupportCategory? = null,
        context: SupportContext? = null
    ): Flow<Result<SupportResponse>> = flow {
        val defaultContext = context ?: createDefaultContext()
        val request = SupportRequest(
            question = question,
            userId = userId,
            context = defaultContext,
            category = category
        )
        
        emit(apiClient.askQuestion(request))
    }
    
    fun getPsychologicalSupportFlow(
        question: String,
        ageGroup: String? = null,
        emotionalState: String? = null
    ): Flow<Result<SupportResponse>> = flow {
        val context = SupportContext(
            currentScreen = null,
            userAgeGroup = ageGroup,
            emotionalState = emotionalState,
            deviceInfo = createDeviceInfo(),
            sessionId = sessionId
        )
        
        val request = SupportRequest(
            question = question,
            userId = userId,
            context = context,
            category = SupportCategory.PSYCHOLOGY
        )
        
        emit(apiClient.getPsychologicalSupport(request))
    }
    
    private fun createDefaultContext(): SupportContext {
        return SupportContext(
            currentScreen = null,
            userAgeGroup = null,
            emotionalState = null,
            deviceInfo = createDeviceInfo(),
            sessionId = sessionId
        )
    }
    
    private fun createDeviceInfo(): DeviceInfo {
        return DeviceInfo(
            platform = "Android",
            version = Build.VERSION.RELEASE,
            model = Build.MODEL,
            screenSize = "unknown",
            orientation = "portrait"
        )
    }
}

